File: iEngine-Web/src/views/cameras/PerspectiveCamera.ts
import { Camera } from "./Camera";

export class PerspectiveCamera extends Camera {
    // 透视参数
    public fov: number = 45; // 视野角度（度）
    public aspect: number = 1; // 宽高比
    public near: number = 0.1; // 近平面
    public far: number = 100; // 远平面
    
    constructor(fov: number = 45, aspect: number = 1, near: number = 0.1, far: number = 100) {
        super();
        this.fov = fov;
        this.aspect = aspect;
        this.near = near;
        this.far = far;
        this.updateProjectionMatrix();
    }
    
    lookAt(x: number, y: number, z: number): void {
        // this.target = new Vector3(x, y, z);
        // this.updateViewMatrix();
        // 上边的方法，会导致“第一人称控制器”无法发挥作用
        this.setTarget(x, y, z);
    }


     // 设置视野角度
    setFov(fov: number): this {
        this.fov = fov;
        this.projectionMatrixDirty = true;
        return this;
    }
    
    // 设置宽高比
    setAspect(aspect: number): this {
        this.aspect = aspect;
        this.projectionMatrixDirty = true;
        return this;
    }
    
    // 设置近平面
    setNear(near: number): this {
        this.near = near;
        this.projectionMatrixDirty = true;
        return this;
    }
    
    // 设置远平面
    setFar(far: number): this {
        this.far = far;
        this.projectionMatrixDirty = true;
        return this;
    }

    // 实现投影矩阵更新
    updateProjectionMatrix(): void {
        if (!this.projectionMatrixDirty) return;
        
        this.projectionMatrix.perspective(
            this.fov * Math.PI / 180, // 转换为弧度
            this.aspect,
            this.near,
            this.far
        );
        
        this.projectionMatrixDirty = false;
    }
    
}

File: iEngine-Native/Engine/src/views/cameras/PerspectiveCamera.cpp
#include "../../../include/iengine/views/cameras/PerspectiveCamera.h"
#include <cmath>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

namespace iengine {
    PerspectiveCamera::PerspectiveCamera(float fov, float aspect, float near, float far)
        : fov(fov), aspect(aspect), near(near), far(far) {
        updateProjectionMatrix();
    }
    
    void PerspectiveCamera::lookAt(float x, float y, float z) {
        setTarget(x, y, z);
    }
    
    void PerspectiveCamera::setFov(float fov) {
        this->fov = fov;
        projectionMatrixDirty_ = true;
    }
    
    void PerspectiveCamera::setAspect(float aspect) {
        this->aspect = aspect;
        projectionMatrixDirty_ = true;
    }
    
    void PerspectiveCamera::setNear(float near) {
        this->near = near;
        projectionMatrixDirty_ = true;
    }
    
    void PerspectiveCamera::setFar(float far) {
        this->far = far;
        projectionMatrixDirty_ = true;
    }
    
    void PerspectiveCamera::updateProjectionMatrix() {
        if (!projectionMatrixDirty_) return;
        
        projectionMatrix_.perspective(
            fov * static_cast<float>(M_PI) / 180.0f, // 转换为弧度
            aspect,
            near,
            far
        );
        
        projectionMatrixDirty_ = false;
    }
    
    const Matrix4& PerspectiveCamera::getProjectionMatrix() {
        updateProjectionMatrix();
        return projectionMatrix_;
    }
}